enterprise/backend/src/metabase_enterprise/data_editing/scope.clj

```diff
@@ -0,0 +1,77 @@
+(ns metabase-enterprise.data-editing.scope
+  (:require
+   [metabase-enterprise.data-editing.types :as types]
+   [metabase.util.malli :as mu]
+   [toucan2.core :as t2]))
+
+;; TODO Perhaps we will prefer the FE just sending this explicitly instead?
+;; For now, it is not doing this, so we infer it.
+(mu/defn scope-type
+  "Classify the scope, useful for switching on in logic and templates."
+  [scope :- ::types/scope.raw] :- :keyword
+  (condp contains? scope
+    :dashcard-id :dashcard
+    :dashboard-id :dashboard
+    :card-id :card
+    :table-id :table
+    :webhook-id :webhook))
+
+(defn- hydrate-from-card-id* [scope]
+  (if (and (contains? scope :collection-id) (contains? scope :table-id))
+    scope
+    (let [card         (t2/select-one [:model/Card :dataset_query :collection_id] (:card-id scope))
+          source-table (-> card :dataset_query :query :source-table)
+          table-id     (when (pos-int? source-table)
+                         source-table)]
+      (-> scope
+          (update :table-id #(or % table-id))
+          (update :collection-id #(or % (:collection_id card)))))))
```

Comment:
maybe `(merge {:table-id 1 :collection-id 2} scope)` then?

Reply:
Yeah that looks nicer. Typically I'm using `or` to avoid querying in the other cases, but these are both free.